db/dohsql.c

```diff
@@ -1032,6 +1032,11 @@ static int _shard_connect(struct sqlclntstate *clnt, dohsql_connector_t *conn,
     conn->clnt->origin = clnt->origin;
     conn->clnt->current_user = clnt->current_user;
     conn->clnt->sql = strdup(sql);
+    /* free parent's authdata if any. it's no longer needed. this also ensures
+     * that get_authdata() returns a new memory which can be freed by
+     * reset_clnt() in _shard_disconnect(). */
+    free(clnt->authdata);
+    clnt->authdata = NULL;
     conn->clnt->authdata = get_authdata(clnt);
```

Comment:
I think doing 
conn->clnt->authdata = clnt->authdata = get_authdata(clnt);
should fix it

Reply:
or we can set clnt->authdata to new alloced memory inside newsql_get_authdata only so that we don't have another bug like this.